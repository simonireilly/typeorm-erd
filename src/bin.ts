#!/usr/bin/env node

import { ERDBuilder } from "./index";
import { resolve } from "path";

const main = async (path: string = "./examples/basic/src/data-source") => {
  const filePath = resolve(path);
  console.log("CWD Stats:", { cwd: process.cwd(), filePath });

  const moduleA = await import(filePath);
  console.info("Module", moduleA);
  const erd = new ERDBuilder("mermaid", moduleA);
  await erd.initialize();
  const erdText = await erd.render();

  console.info(erdText);
};

main();
