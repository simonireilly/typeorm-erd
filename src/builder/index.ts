import { DataSource, EntityMetadata, EntitySchema } from "typeorm";
import { ConnectionMetadataBuilder } from "typeorm/connection/ConnectionMetadataBuilder";
import { RelationType } from "typeorm/metadata/types/RelationTypes";

export type BuilderRelations = Record<
  string,
  {
    entityRelations: RelationData[];
  }
>;

interface RelationData {
  relationType: RelationType;
  propertyPath: string;
  nullable: boolean;
  isOwning: boolean;
  inverseSidePropertyPath: string;
  source: string;
  target: string;
}

const entityMetaData = async (connection: DataSource) => {
  const connectionMetadataBuilder = new ConnectionMetadataBuilder(connection);

  const { entities } = connection.options;
  const TEntities = entities as (Function | EntitySchema<any> | string)[];

  let entityMetadata;

  if (entities) {
    entityMetadata = await connectionMetadataBuilder.buildEntityMetadatas(
      TEntities
    );
  } else {
    throw Error("No entities found on connection");
  }

  return entityMetadata;
};

/**
 * Return the relation data for each table
 *
 */
const relations = (meta: EntityMetadata[]): BuilderRelations => {
  const relationData = meta.reduce<BuilderRelations>((acc, entity) => {
    const entityRelations = entity.ownRelations.map(
      ({
        relationType,
        inverseEntityMetadata,
        propertyPath,
        inverseSidePropertyPath,
        isOwning,
      }) => {
        return {
          relationType,
          propertyPath,
          isOwning,
          nullable:
            entity.columns.find((c) => c.propertyName === propertyPath)
              ?.isNullable || false,
          inverseSidePropertyPath,
          source: entity.tableName,
          target: inverseEntityMetadata.tableName,
        };
      }
    );

    return {
      ...acc,
      [entity.tableName]: {
        entityRelations,
      },
    };
  }, {});

  return relationData;
};

export const builders = {
  entityMetaData,
  relations,
};
