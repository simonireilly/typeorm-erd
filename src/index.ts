export * from "./builder";
export * from "./adapters";
