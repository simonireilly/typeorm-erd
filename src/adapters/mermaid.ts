import { DataSource, EntityMetadata } from "typeorm";
import { RelationMetadata } from "typeorm/metadata/RelationMetadata";
import { BuilderRelations, builders } from "../builder";

const RelationShips = {
  left: {
    "one-to-many": "||",
    "many-to-one": "}|",
    "one-to-one": "||",
    "many-to-many": "}|",
  },
  right: {
    "one-to-many": "|{",
    "many-to-one": "||",
    "one-to-one": "||",
    "many-to-many": "|{",
  },
};

export class Mermaid {
  private meta: EntityMetadata[];
  private relations: BuilderRelations;

  constructor(
    readonly dataSource: DataSource,
    private entityBuilder = builders.entityMetaData,
    private relationBuilder = builders.relations
  ) {
    this.dataSource = dataSource;
    this.entityBuilder = entityBuilder;
  }

  public async initialize() {
    this.meta = await this.entityBuilder(this.dataSource);
    this.relations = await this.relationBuilder(this.meta);
  }

  public render(): string {
    return `erDiagram\n  ${this.renderTables()}\n  ${this.renderRelations()}`;
  }

  public renderRelations() {
    return this.buildRelations().join("\n  ");
  }

  public buildRelations() {
    return Object.entries(this.relations).reduce((acc, [key, values]) => {
      const relations = values.entityRelations.map((rel) => {
        // Remove unnamed relations and only add relations for the explicit owner
        if (!rel.propertyPath || !rel.isOwning) return "";

        return `${rel.source} ${RelationShips["left"][rel.relationType]}--${
          RelationShips["right"][rel.relationType]
        } ${rel.target}: ${rel.propertyPath}`;
      });

      return [...acc, ...relations.filter(Boolean)];
    }, [] as string[]);
  }

  private renderTables() {
    return this.meta
      .map((entry) => {
        const columns = entry.columns.map((column) => {
          return [
            this.dataSource.driver.normalizeType(column),
            column.databaseName,
            column.isPrimary ? "PK" : column.referencedColumn ? "FK" : "",
            column?.comment && `"${column?.comment}"`,
          ]
            .filter(Boolean)
            .join(" ");
        });

        return `${entry.tableName} {\n    ${columns.join("\n    ")}\n  }`;
      })
      .join("\n  ");
  }
}
