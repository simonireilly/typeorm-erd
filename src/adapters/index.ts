import { DataSource } from "typeorm";
import { MermaidErd } from "./mermaid";
import { PlantUMLErd } from "./plantuml";

const BUILDERS = {
  mermaid: MermaidErd,
  plantuml: PlantUMLErd,
};

type TBuildTypes = keyof typeof BUILDERS;

class ERDBuilder {
  private instance: MermaidErd | PlantUMLErd;

  constructor(public instanceType: TBuildTypes, dataSource: DataSource) {
    const adapter = BUILDERS[instanceType];
    this.instance = new adapter(dataSource);
  }

  public async initialize() {
    return this.instance.initialize();
  }

  public render(): string {
    return this.instance.render();
  }
}

export { PlantUMLErd, MermaidErd, ERDBuilder };
