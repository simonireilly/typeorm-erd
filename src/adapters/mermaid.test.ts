import { AppDataSource } from "../../examples/basic/src/data-source";
import { AppDataSource as CartData } from "../../examples/shopping-cart/src/data-source";
import { Mermaid } from "./mermaid";

describe("mermaid adapter", () => {
  describe("integration tests", () => {
    test("renders the complete basic tables", async () => {
      const erd = new Mermaid(AppDataSource);
      await erd.initialize();

      expect(erd.render()).toMatchSnapshot();
    });
    test("renders the complete cart tables", async () => {
      const erd = new Mermaid(CartData);
      await erd.initialize();

      expect(erd.render()).toMatchSnapshot();
    });
  });

  describe("public methods", () => {
    it("renderRelations", async () => {
      const erd = new Mermaid(AppDataSource);

      await erd.initialize();

      expect(erd.buildRelations()).toEqual(["comment }|--|| user: author"]);
    });
  });
});
