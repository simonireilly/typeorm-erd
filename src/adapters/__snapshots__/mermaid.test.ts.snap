// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mermaid adapter integration tests renders the complete basic tables 1`] = `
"erDiagram
  user {
    integer id PK \\"The primary user identifier\\"
    varchar firstName
    varchar lastName
    integer age
  }
  comment {
    integer id PK \\"The ID for the comment\\"
    varchar text
    integer likes
    integer authorId FK \\"The primary user identifier\\"
  }
  comment }|--|| user: author"
`;

exports[`mermaid adapter integration tests renders the complete cart tables 1`] = `
"erDiagram
  cart_item {
    bigint id PK
    bigint productId FK
    bigint cartId FK
    varchar sku
    float price
    float discount
    smallint quantity
    tinyint active
    datetime createdAt
    datetime updatedAt
    text content
  }
  category {
    bigint id PK
    bigint parentId FK
    varchar title
    varchar metaTitle
    varchar slug
    text content
  }
  product_meta {
    bigint id PK
    bigint productId FK
    varchar key
    text content
  }
  product_review {
    bigint id PK
    bigint productId FK
    bigint parentId FK
    varchar title
    smallint rating
    tinyint published
    datetime createdAt
    datetime publishedAt
    text content
  }
  tag {
    bigint id PK
    varchar title
    varchar metaTitle
    varchar slug
    text content
  }
  product {
    bigint id PK
    bigint userId FK
    varchar title
    varchar metaTitle
    varchar slug
    tinytext summary
    smallint type
    varchar sku
    float price
    float discount
    smallint quantity
    tinyint shop
    datetime createdAt
    datetime updatedAt
    datetime publishedAt
    datetime startsAt
    datetime endsAt
    text content
  }
  order_item {
    bigint id PK
    bigint productId FK
    bigint orderId FK
    varchar sku
    float price
    float discount
    smallint quantity
    datetime createdAt
    datetime updatedAt
    text content
  }
  transaction {
    bigint id PK
    bigint userId FK
    bigint orderId FK
    varchar code
    smallint type
    smallint mode
    smallint status
    datetime createdAt
    datetime updatedAt
    text content
  }
  order {
    bigint id PK
    bigint userId FK
    varchar sessionId
    varchar token
    smallint status
    float subTotal
    float itemDiscount
    float tax
    float shipping
    float total
    varchar promo
    float discount
    float grandTotal
    varchar firstName
    varchar middleName
    varchar lastName
    varchar mobile
    varchar email
    varchar line1
    varchar line2
    varchar city
    varchar province
    varchar country
    datetime createdAt
    datetime updatedAt
    text content
  }
  user {
    bigint id PK
    enum role
    varchar firstName
    varchar middleName
    varchar lastName
    varchar mobile
    varchar email
    varchar passwordHash
    tinyint admin
    tinyint vendor
    datetime registeredAt
    datetime lastLogin
    tinytext intro
    text profile
  }
  cart {
    bigint id PK
    bigint userId FK
    varchar sessionId
    varchar token
    smallint status
    varchar firstName
    varchar middleName
    varchar lastName
    varchar mobile
    varchar email
    varchar line1
    varchar line2
    varchar city
    varchar province
    varchar country
    datetime createdAt
    datetime updatedAt
    text content
  }
  product_category {
    bigint categoryId PK
    bigint productId PK
  }
  product_tag {
    bigint productId PK
    bigint tagId PK
  }
  cart_item }|--|| cart: cart
  cart_item }|--|| product: product
  category }|--|| category: parent
  category }|--|{ product: products
  product_meta }|--|| product: product
  product_review }|--|| product_review: parent
  product_review }|--|| product: product
  product }|--|| user: user
  product }|--|{ tag: tags
  order_item }|--|| order: order
  order_item }|--|| product: product
  transaction }|--|| order: order
  transaction }|--|| user: user
  order }|--|| user: user
  cart }|--|| user: user"
`;
